packages/storybook/src/generators/configuration/lib/ensure-dependencies.ts

```diff
@@ -71,9 +71,16 @@ export function ensureDependencies(
     if (isPnpm) {
       // If it's pnpm, it needs the framework without the builder
       // as a dependency too (eg. @storybook/react)
-      const matchResult = options.uiFramework?.match(/^@storybook\/(\w+)/);
-      const uiFrameworkWithoutBuilder = matchResult ? matchResult[0] : null;
-      if (uiFrameworkWithoutBuilder) {
+      const matchResult = options.uiFramework?.match(
+        /^@storybook\/([\w-]+?)(?:-(?:vite|webpack5|webpack))?$/
```

Comment:
Nit: I don't think `?:` in `(?:vite|webpack5|webpack)` really matters? as we're not checking `matchResult[2]` anyway. But it works either way.